lib/boavizta.ts

```diff
@@ -0,0 +1,283 @@
+}
+
+
+export class BoaviztaCpuImpactModel extends BoaviztaImpactModel implements IImpactModelInterface {
+    private readonly componentType = "cpu";
+    sharedParams: object | undefined = undefined;
+    public name: string | undefined;
+    public verbose: boolean = false;
+    public allocation: string = "TOTAL";
+
+    constructor() {
+        super();
+        this.metricType = "cpu";
+        this.componentType = "cpu";
+    }
+
+    modelIdentifier(): string {
+        return "org.boavizta.cpu.sci"
+    }
+
+    protected async captureStaticParams(staticParams: object): Promise<object> {
+        if ('verbose' in staticParams) {
+            this.verbose = staticParams.verbose as boolean ?? false;
+            staticParams.verbose = undefined;
+        }
+        if ('allocation' in staticParams) {
+            const allocation = staticParams.allocation as string ?? "TOTAL";
+            if (["TOTAL", "LINEAR"].includes(allocation)) {
+                this.allocation = allocation;
+            } else {
+                throw new Error("Improper configure: Invalid allocation parameter. Either TOTAL or LINEAR");
+            }
+            staticParams.allocation = undefined;
+        }
+        if (!('name' in staticParams)) {
+            throw new Error("Improper configure: Missing name parameter");
+        }
+        if (!('core_units' in staticParams)) {
+            throw new Error("Improper configure: Missing core_units parameter");
+        }
+        this.sharedParams = Object.assign({}, staticParams);
+        return this.sharedParams
+    }
+
+
+    async fetchData(usageData: object | undefined): Promise<object> {
+        if (this.sharedParams === undefined) {
+            throw new Error("Improper configure: Missing configuration parameters")
+        }
+        const dataCast = this.sharedParams as { [key: string]: any };
+        dataCast['usage'] = usageData
+        const response = await axios.post(`https://api.boavizta.org/v1/component/${this.componentType}?verbose=${this.verbose}&allocation=${this.allocation}`, dataCast);
+        return this.formatResponse(response);
+    }
+}
+
+export class BoaviztaCloudImpactModel extends BoaviztaImpactModel implements IImpactModelInterface {
+    public sharedParams: object | undefined = undefined;
+    public instanceTypes: { [key: string]: string[] } = {};
+    public name: string | undefined;
+    public verbose: boolean = false;
+    public allocation: string = "TOTAL";
+
+    modelIdentifier(): string {
+        return "org.boavizta.cloud.sci"
+    }
+
+    protected async captureStaticParams(staticParams: object) {
+        const staticParamsCast = staticParams;
+        if ('verbose' in staticParamsCast) {
+            this.verbose = staticParamsCast.verbose as boolean ?? false;
+            staticParamsCast.verbose = undefined;
+        }
+        if ('allocation' in staticParamsCast) {
+            const allocation = staticParamsCast.allocation as string ?? "TOTAL";
```

Comment:
I'm not aware of this syntax, what does ?? do here?

Reply:
null coalescing operator. it returns the second value if the first is undefined or null.